File: LightningConstructs/Source/Socket/Properties.cs
namespace Lightning
{
    public partial class Socket
    {
        public System.Int32 Available => NetSocket.Available;
        public System.Boolean Listening => IsListening;
        public System.Boolean Connected
        {
            get
            {
                if (IsConnected == false) { return false; }

                lock (ReadLock)
                {
                    lock (WriteLock)
                    {
                        var DataAvailable = NetSocket.Poll(0, System.Net.Sockets.SelectMode.SelectRead);
                        var DataNotAvailable = (NetSocket.Available == 0);
                
                        //A - DataAvailable, B - DataNotAvailable
                        //
                        //A = true,  B = false - connected, data is available for reading. not inverts to true
                        //A = false, B = true  - connected, data is not available for reading. not inverts to true
                        //A = false, B = false - connected, although this value should be impossible to reach. not inverts to true
                        //A = true,  B = true  - disconnected, data is not available for reading. not inverts to false
                
                        IsConnected = !(DataAvailable == true && DataNotAvailable == true);
                        return IsConnected;
                    }
                }
            }
        }
    }
}

File: LightningConstructs/Source/Socket/Properties.vb
    Public Partial Class Socket
        Public ReadOnly Property Available As Int32
            Get
                Return NetSocket.Available
            End Get
        End Property

        Public ReadOnly Property Listening As System.Boolean
            Get
                Return IsListening
            End Get
        End Property

        Public ReadOnly Property Connected As System.Boolean
            Get
                If IsConnected = False Then Return False
                SyncLock ReadLock
                    Dim DataAvailable = NetSocket.Poll(0, Net.Sockets.SelectMode.SelectRead)
                    Dim DataNotAvailable = (NetSocket.Available = 0)
                    IsConnected = Not (DataAvailable = True AndAlso DataNotAvailable = True)
                    Return IsConnected
                End SyncLock
            End Get
        End Property
    End Class
